File: backend/app/modules/commcalc/rep_incentive_range.py
"""REP INCENTIVE — A MONTH RANGE ON ONE PAGE (owner 2026-09-25).

Owner, verbatim: *"al for rep incentive report the range for multiple months should be there to display the
commission for teh months selected in different rowqn on one page"*.

WHAT THIS IS — AND IS NOT. The Rep Incentive report (`/commcalc/reports`) reads ONE month through
`GET /commcalc/commissions/{period}` (`router.get_commissions`: the stored `rep_commissions` rows, the
chargeback and ops-chargeback deductions, the market stamp, the caller's self / span scope). A range is that
SAME per-month read, LOOPED over the months `account/_period.month_range` enumerates — never a second
calculation. This module only lays the per-month answers side by side: one row per rep per month, a
subtotal per month, a total per rep across the range, and a grand total. Every number is copied from the
per-month payload; the only arithmetic is summing, rounded to the cent. PURE (no I/O).

The cap is `MAX_MONTHS` = 12: each month is one full per-month read (a rep_commissions query, the
chargeback reads and the scope resolution), so a year is the widest window that stays one page.
"""

MAX_MONTHS = 12

# the money fields a range sums (every one is a field the single-month payload already carries)
MONEY_FIELDS = ("subtotal", "total_payout", "chargeback_deduction", "ops_chargeback_deduction", "final_payout",
                "plan_comm", "installment_comm_sale", "residual_installment_comm", "acc_comm")
COUNT_FIELDS = ("premium_acts", "byod_acts", "upgrade_acts")


def _f(v):
    try:
        return float(v or 0)
    except (TypeError, ValueError):
        return 0.0


def _r2(x):
    return round(x + 0.0, 2)


def rep_key(row):
    """The rep a row belongs to across months — the roster name when present, else the POS name, upper."""
    r = row or {}
    return (str(r.get("storeops_name") or "").strip() or str(r.get("epay_salesperson") or "").strip()).upper()


def assemble(months, per_month):
    """`months` = the canonical month list (oldest first); `per_month` = {month: [row, …]} — each list
    EXACTLY what the single-month read returned for that month. Returns

      {"months": [...],
       "rows": [row + {"range_month": month}, …]     — month order, then the per-month order kept,
       "month_totals": [{"period", <money>…, <counts>…, "reps"}, …],
       "rep_totals": [{"rep", "store", "months", <money>…, <counts>…}, …]   — biggest payout first,
       "grand_total": {<money>…, <counts>…, "reps", "months_with_rows"}}

    Rows are COPIED (a shallow dict copy + a `range_month` key); no field of a per-month row changes."""
    rows, month_totals, reps = [], [], {}
    grand = {k: 0.0 for k in MONEY_FIELDS}
    grand.update({k: 0 for k in COUNT_FIELDS})
    with_rows = 0
    for m in months:
        mrows = list(per_month.get(m) or [])
        if mrows:
            with_rows += 1
        mt = {"period": m, "reps": len(mrows)}
        for k in MONEY_FIELDS:
            mt[k] = _r2(sum(_f(r.get(k)) for r in mrows))
        for k in COUNT_FIELDS:
            mt[k] = int(sum(int(_f(r.get(k))) for r in mrows))
        month_totals.append(mt)
        for r in mrows:
            d = dict(r)
            d["range_month"] = m
            rows.append(d)
            key = rep_key(r)
            t = reps.setdefault(key, {"rep": r.get("storeops_name") or r.get("epay_salesperson"),
                                      "store": r.get("store"), "months": 0,
                                      **{k: 0.0 for k in MONEY_FIELDS}, **{k: 0 for k in COUNT_FIELDS}})
            t["months"] += 1
            for k in MONEY_FIELDS:
                t[k] = _r2(t[k] + _f(r.get(k)))
            for k in COUNT_FIELDS:
                t[k] += int(_f(r.get(k)))
        for k in MONEY_FIELDS:
            grand[k] = _r2(grand[k] + mt[k])
        for k in COUNT_FIELDS:
            grand[k] += mt[k]
    grand["reps"] = len(reps)
    grand["months_with_rows"] = with_rows
    return {"months": list(months), "rows": rows, "month_totals": month_totals,
            "rep_totals": sorted(reps.values(), key=lambda t: (-t["total_payout"], str(t["rep"] or ""))),
            "grand_total": grand, "max_months": MAX_MONTHS}

File: frontend/src/app/(platform)/commcalc/_lib/repIncentiveRange.ts
// REP INCENTIVE — A MONTH RANGE ON ONE PAGE (owner 2026-09-25): "the range for multiple months should be
// there to display the commission for teh months selected in different rowqn on one page".
//
// The rows come from GET /commcalc/commissions-range, which LOOPS the single-month report's own handler
// (get_commissions) over the months account/_period.month_range enumerates — so every rep row here IS the
// row that month shows alone. This module is PURE and DISPLAY-ONLY: it groups those rows by month for the
// table and sums the ones on screen (the single-month page sums its filtered rows the same way). It
// imports nothing, so frontend/tools/rep-incentive-range-proof.mjs transpiles and proves it directly.

/** The widest range the page asks for — the backend's rep_incentive_range.MAX_MONTHS (12). */
export const RANGE_MAX_MONTHS = 12

export type RangeRow = { range_month: string; [k: string]: any }

export const RANGE_MONEY = ['subtotal', 'total_payout', 'final_payout'] as const
export const RANGE_COUNTS = ['premium_acts', 'byod_acts', 'upgrade_acts'] as const

const num = (v: any): number => { const n = Number(v); return Number.isFinite(n) ? n : 0 }
const cents = (n: number): number => Math.round((n + Number.EPSILON) * 100) / 100

export type MonthBlock = { month: string; rows: RangeRow[]; totals: Record<string, number> }

function totalsOf(rows: RangeRow[]): Record<string, number> {
  const t: Record<string, number> = { reps: rows.length }
  for (const k of RANGE_MONEY) t[k] = cents(rows.reduce((s, r) => s + num(r[k]), 0))
  for (const k of RANGE_COUNTS) t[k] = rows.reduce((s, r) => s + Math.trunc(num(r[k])), 0)
  return t
}

/** One block per month in `months` order (a month with no rows still shows, empty), each with the
 *  rows the backend returned for it — order kept — and their sums; plus the grand total. PURE. */
export function monthBlocks(rows: RangeRow[], months: string[]): { blocks: MonthBlock[]; grand: Record<string, number> } {
  const by = new Map<string, RangeRow[]>()
  for (const m of months || []) by.set(m, [])
  for (const r of rows || []) {
    const m = String(r?.range_month ?? '')
    if (!by.has(m)) by.set(m, [])
    by.get(m)!.push(r)
  }
  const blocks = Array.from(by.entries()).map(([month, rs]) => ({ month, rows: rs, totals: totalsOf(rs) }))
  return { blocks, grand: totalsOf(blocks.flatMap(b => b.rows)) }
}

/** 'YYYY-MM' of today's month minus `back` months — the default window's bounds. PURE given `today`. */
export function monthInput(today: Date, back = 0): string {
  const d = new Date(today.getFullYear(), today.getMonth() - back, 1)
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}`
}

/** How many months [from, to] spans (inclusive); 0 when either bound is not 'YYYY-MM' or reversed. */
export function spanMonths(from: string, to: string): number {
  const p = (s: string) => { const m = /^(\d{4})-(\d{2})$/.exec(String(s || '')); return m ? [Number(m[1]), Number(m[2])] : null }
  const a = p(from), b = p(to)
  if (!a || !b) return 0
  const n = (b[0] - a[0]) * 12 + (b[1] - a[1]) + 1
  return n > 0 ? n : 0
}
